registry/storage/registry.go

```diff
@@ -132,6 +135,85 @@ func (reg *registry) Repository(ctx context.Context, name string) (distribution.
 	}, nil
 }
 
+// listScope recursively walks directories under given scopePath and returns a
+// list of all repository names found. scope is a suffix of scopePath that
+// starts a registry name.
+func (reg *registry) listScope(ctx context.Context, scopePath, scope string) (*list.List, error) {
+	d := reg.blobStore.driver
+	rl := list.New()
+	isRepository, err := exists(ctx, d, path.Join(scopePath, manifestsDirectory))
+	if err != nil {
+		return nil, err
+	}
+	if isRepository {
+		// Assume that repository doesn't contain sub-repositories
+		rl.PushBack(scope)
+		return rl, nil
+	}
+	logger := context.GetLogger(ctx)
+	children, err := d.List(ctx, scopePath)
+	if err != nil {
+		return nil, err
+	}
+	for _, child := range children {
+		name := path.Base(child)
+		subScope := path.Join(scope, name)
+		info, err := d.Stat(ctx, child)
+		if err != nil {
+			logger.Warnf("Failed to stat scope %q: %v", subScope)
+			continue
+		}
+		if !info.IsDir() {
+			continue
+		}
+		if !v2.RepositoryNameComponentAnchoredRegexp.Match([]byte(name)) || len(subScope) > v2.RepositoryNameTotalLengthMax {
+			logger.Debugf("Skipping invalid scope %q", child)
+			continue
+		}
+		scopeList, err := reg.listScope(ctx, child, subScope)
+		if err != nil {
+			logger.Warnf("Failed to list scope %q: %v", subScope, err)
+			continue
+		}
+		rl.PushBackList(scopeList)
+	}
+	return rl, nil
+}
+
+// listRepositories list of all repository names found under registry root.
+func (reg *registry) listRepositories(ctx context.Context) (*list.List, error) {
```

Comment:
> Methods should never return list.List as a type. Please use slices so the contents of the list is clear.

I just wanted to avoid allocation of unnecessarily large arrays multiple times or using some custom allocation heuristics. But I see the importance of type checking and visibility. I'll remove lists from function declarations.
> Also, this should be covered by catalog calls.

Ahh! Missed that one. Thanks for pointing this out.